pkg/armrpc/asyncoperation/worker/service.go

```diff
@@ -18,57 +18,64 @@ package worker
 
 import (
 	"context"
+	"sync"
 
 	manager "github.com/radius-project/radius/pkg/armrpc/asyncoperation/statusmanager"
-	"github.com/radius-project/radius/pkg/armrpc/hostoptions"
 	"github.com/radius-project/radius/pkg/ucp/dataprovider"
-	queue "github.com/radius-project/radius/pkg/ucp/queue/client"
 	qprovider "github.com/radius-project/radius/pkg/ucp/queue/provider"
 	"github.com/radius-project/radius/pkg/ucp/ucplog"
 )
 
 // Service is the base worker service implementation to initialize and start worker.
+// All exported fields should be initialized by the caller.
 type Service struct {
-	// ProviderName is the name of provider namespace.
-	ProviderName string
-	// Options is the server hosting options.
-	Options hostoptions.HostOptions
-	// StorageProvider is the provider of storage client.
-	StorageProvider dataprovider.DataStorageProvider
 	// OperationStatusManager is the manager of the operation status.
 	OperationStatusManager manager.StatusManager
-	// Controllers is the registry of the async operation controllers.
-	Controllers *ControllerRegistry
-	// RequestQueue is the queue client for async operation request message.
-	RequestQueue queue.Client
+
+	// Options configures options for the async worker.
+	Options Options
+
+	// QueueProvider is the queue provider. Will be initialized from config if not provided.
+	QueueProvider *qprovider.QueueProvider
+
+	// StorageProvider is the provider of storage client. Will be initialized from config if not provided.
+	StorageProvider dataprovider.DataStorageProvider
+
+	// controllers is the registry of the async operation controllers.
+	controllers *ControllerRegistry
+
+	// controllersInit is used to ensure single initialization of controllers.
+	controllersInit sync.Once
 }
 
-// Init initializes worker service - it initializes the StorageProvider, RequestQueue, OperationStatusManager, Controllers, KubeClient and
-// returns an error if any of these operations fail.
-func (s *Service) Init(ctx context.Context) error {
-	s.StorageProvider = dataprovider.NewStorageProvider(s.Options.Config.StorageProvider)
-	qp := qprovider.New(s.Options.Config.QueueProvider)
-	var err error
-	s.RequestQueue, err = qp.GetClient(ctx)
-	if err != nil {
-		return err
-	}
-	s.OperationStatusManager = manager.New(s.StorageProvider, s.RequestQueue, s.Options.Config.Env.RoleLocation)
-	s.Controllers = NewControllerRegistry(s.StorageProvider)
-	return nil
+// Controllers returns the controller registry.
+func (s *Service) Controllers() *ControllerRegistry {
+	s.controllersInit.Do(func() {
+		s.controllers = NewControllerRegistry(s.StorageProvider)
+	})
+
+	return s.controllers
 }
```

Comment:
Are we relying on this function to be called only once? I am asking because we are not checking if `s.controllers` already has some value. Do we need to check that first?

Reply:
This is what `sync.Once` does for us. https://pkg.go.dev/sync#Once.Do